### ISSUE TITLE
Running example code got error in Windows 10

### ISSUE THREAD

**Issue by rangerbottle** [dotnet/TorchSharp#1041]
Dear experts,

Got this error during running TorchSharp example code

"
[W C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\torch\csrc\jit\codegen\cuda\interface.cpp:47] Warning: Loading nvfuser library failed with: error in LoadLibrary for nvfuser_codegen.dll. WinError 126: The specified module could not be found.
 (function LoadingNvfuserLibrary)
"

I would like ask your help.
OS: win10
Pytorch 1.12.1+cu116
CUDA 11.6

Thanks

**Comment by dje-dev** [comment]
Yes, I was just about to report that I'm getting exactly this same error. Windows 10, latest version of Torchsharp, CUDA 11.7.

**Comment by NiklasGustafsson** [comment]
The current version of TorchSharp requires CUDA 11.7-compatible drivers. You don't need the CUDA toolkit installed, but the drivers have to be up-to-date.

Which example are you running? From this repo, or from the TorchSharpExamples repo?

**Comment by dje-dev** [comment]
I run some custom software, written in C# against TorchSharp (which works great, thank you!). This error seems to only happen when I load a torchscript file and then invoke forward on a module. Actually the TorchScript module does work correctly, but runs at only about 70% of the  expected speed (compared to native Pytorch on the same system). I don't think this is TorchSharp overhead, since the network is large and there would be few transitions between C# and the torch libraries.  I see the nvfuser.dll is present so apparently it is expected that this should have been accessible.

**Comment by NiklasGustafsson** [comment]
What NuGet packages are you using?

**Comment by NiklasGustafsson** [comment]
Oh, I see. Does the TorchScript model use TorchVision? There are a couple of TV operators that require native support that we don't have in TorchSharp. I'll try to dig up the other thread on this.

**Comment by NiklasGustafsson** [comment]
I wonder if this could help:

https://github.com/dotnet/TorchSharp/issues/1003

**Comment by rangerbottle** [comment]
I used demo code from this repo
        static async Task Main(string[] args)
        {
            var lin1 = Linear(1000, 100);
            var lin2 = Linear(100, 10);
            var seq = Sequential(("lin1", lin1), ("relu1", ReLU()), ("drop1", Dropout(0.1)), ("lin2", lin2));

            var x = torch.randn(64, 1000);
            var y = torch.randn(64, 10);

            var optimizer = torch.optim.Adam(seq.parameters());

            for (int i = 0; i < 10; i++)
            {
                var eval = seq.forward(x);
                var output = functional.mse_loss(eval, y, Reduction.Sum);

                optimizer.zero_grad();

                output.backward();

                optimizer.step();
            }

And my nuget package of TorchSharp is 0.100.3
NVIDIA-SMI 516.94  and Driver Version: 516.94
Currently my CUDA version v11.6
So I should upgrade my GPU driver?
Thanks


> The current version of TorchSharp requires CUDA 11.7-compatible drivers. You don't need the CUDA toolkit installed, but the drivers have to be up-to-date.
> 
> Which example are you running? From this repo, or from the TorchSharpExamples repo?

**Comment by NiklasGustafsson** [comment]
Do you also have a backend package? The libtorch-* packages, is what I mean? If not, the easiest thing to do is to include `TorchSharp-cuda-windows` instead of `TorchSharp`.

**Comment by BowserKingKoopa** [comment]
I'm experiencing the same thing. It happens when I run anything:
```
        static void Main(string[] args)
        {
            var t = torch.zeros(10);
            Console.ReadLine();
        }
```
I'm using TorchSharp-cuda-windows (0.100.3)
NVIDIA Geforce RTX 3090
Drivers 536.25
CUDA V12.2.91
.NET 6
Windows 10

**Comment by NiklasGustafsson** [comment]
Does it work with TorchSharp-cpu?

Can you try with this tutorial from the Examples repo: https://github.com/dotnet/TorchSharpExamples/blob/main/tutorials/CSharp/tutorial5.ipynb

(You'll need .NET Interactive and VS Code to run the notebook)

**Comment by BowserKingKoopa** [comment]
> Does it work with TorchSharp-cpu?

Yes. I do not get the error with TorchSharp-cpu. Which I expected because the error message seems CUDA related.

**Comment by NiklasGustafsson** [comment]
When you say that you have CUDA 12.2..., do you mean that you have the CUDA toolkit installed? It shouldn't have any effect on libtorch, which bundles the necessary CUDA binaries, but I'm just curious.

**Comment by BowserKingKoopa** [comment]
> When you say that you have CUDA 12.2..., do you mean that you have the CUDA toolkit installed? It shouldn't have any effect on libtorch, which bundles the necessary CUDA binaries, but I'm just curious.

Yes, that's what I meant. The NVIDIA CUDA toolkit.

**Comment by NiklasGustafsson** [comment]
Are you seeing an actual error, or just the warning?

**Comment by rangerbottle** [comment]
> 

![image](https://github.com/dotnet/TorchSharp/assets/15903780/92a12a8c-f89a-450d-8ea6-9bb31ead074e)
Yes, I use latest Torchsharp-cuda-windows

**Comment by miroslav22** [comment]
I've just been tearing my hair out with this issue... and the solution doesn't seem to make a lot of sense.  If you add to the top of your program:

var device = torch.cuda_is_available() ? torch.device(DeviceType.CUDA) : torch.device(DeviceType.CPU);

Then you don't get the error...  Note that "torch.cuda_is_available()" seems to be key.  If you just call "torch.device(DeviceType.CUDA)" then it still fails.

Hope it helps someone else anyway.

**Comment by NiklasGustafsson** [comment]
@miroslav22 -- this is very useful! Can you try something for me?

Instead of calling `cuda_is_available()`, could you precede the call to torch.device(DeviceType.CUDA) with:

```C#
torch.TryInitializeDeviceType(DeviceType.CUDA);
... torch.device(DeviceType.CUDA) ...
```

**Comment by miroslav22** [comment]
Yes, that also fixes it.  Thanks :)

**Comment by NiklasGustafsson** [comment]
With this change, trying to call `torch.device(DeviceType.CUDA)`, or `torch.cuda()`, etc. will throw an exception if CUDA isn't available.

**Comment by NiklasGustafsson** [comment]
I believe that this problem was fixed in v0.100.5. @rangerbottle, @miroslav22 -- if you could verify that it works for you, and if so, close the issue, that would be great!

**Comment by NiklasGustafsson** [comment]
Haven't heard, so I'll assume that it was fixed (it really should have). I'll close the issue -- if the problem persists, please reopen.
